### PR TITLE
Created event handlers for lobby full and confirm start game and tests.

### LINKED ISSUE #49: Design the application domain model
### Description
Need to design domain model objects which will be used to encapsulate the key details of entities in the Mafia gaming environment. These include:
- The Lobby which represents a session which multiple players can join
    - Properties: `roomName (lobby code), players[], gameState, roundNumber`
- The players joining the game, their details and identifying information
    - Properties: `socketID, nickname, role, isAlive`
- The roles held by the players (mafia, detective, etc.)
    - Enum values: `Mafia, Detective, Medic, Civilian` 
- The different game states which represent the different phases in the game cycle
    - Game states: `Night Time, Day Time, Initial Vote, Trial`
- A container to represent all the ongoing game sessions which is a collection of all the active lobbies
    - Property: `Map(roomID, Room)`

### Sub-Task List
- [x] Create a base domain model for Lobby, Player, Role, Game State and Lobby Container with properties
- [x] Add getter and setter functions
- [x] Add tests to the CI pipeline to test the functionality 

### Issue Approval
- [x]  Issue approved by: Dhors (Daniel Hor), Jteo97 (Justin Teo)

### LINKED ISSUE #87: Create game ready event handler.
### In-Depth Feature Description:
This feature is to let players know the lobby is full and ready to start. It requires the host to be allowed to start the game and informing the non-host players that the game will be starting. This involves sending two different events to the host and non-host players respectively.

This is the last stage of screens before the game begins.

### Sub-Task List:
- [x] Create event handler for the lobby filling up. 
- [x] Create tests to test functionality of event handler.

### Issue Approval:

- [x] Issue approved by: @LindaBot

### LINKED ISSUE #100: 6 empty players are created by the Room upon initialisation, and 7 players can be added instead of 6.
**Describe the bug**
In `room.js`, instead of creating an array of size 6, an array with 6 empty items are created. When we add the first player, the array is now size 7. 

Line 14:
```JavaScript
this.players = new Array(this.maxPlayerCount);
```
Should be
```JavaScript
this.players = new Array();
```

Line 35:
```JavaScript
if (player !== null && this.players.length <= this.maxPlayerCount) {
```
Should be 
```JavaScript
if (player !== null && this.players.length < this.maxPlayerCount) {
```

**To Reproduce**
* Create a new Room and try to add players to it. 

**Expected behavior**
An array should be of created, without 6 empty items and maximum of 6 players should be allowed. 

### Issue Approval:
- [x]  Issue approved by: Dhors

### PR DESCRIPTION
**Pull request by Dhors**
# Description
This PR fixes #100 and fixes #87 and fixes #49

Created event handler for the lobby filling up to maximum capacity, and event handler for the host starting game.

tests have been included to tests these event handlers using a separate event handler ("create-lobby") to initialize the socket object with a player and roomID.

A bug has been found and fixed involving the room classes' players array. This was fixed to not initialize the array with 6 blank objects, and the max size if statement was corrected.

an additional test suite has been created to ensure the domain model classes/objects are functioning correctly.

Before: 
image

image

After: 
image

image

# Type of change:
* New feature (Non-breaking change that adds functionality)
* Bug fix (Non-breaking change which fixes a bug)

### CODE REVIEW

File: backend/Events/LobbyEvents.js [hunk 1]
```diff
@@ -10,23 +10,56 @@ const MafiaGame = require('../domain/MafiaGame');
...
+    mafiaGame.gameRoomsDict[roomID].addPlayer(host);
+
+    // Subscribe to the room events
+    socket.join(roomID);
+
+    // Add player information to the host socket
+    socket.player = host;
+
+    // Send room ID back to host.
+    io.in(roomID).emit("lobby-code", new LobbyCodeDTO(roomID));
```
Comment (UltimaKuma) [inline backend/Events/LobbyEvents.js:27]: Sends to everyone in the room, but should be fine since host will always be the only player in the room at this point

File: backend/Events/LobbyEvents.js [hunk 2]
```diff
@@ -10,23 +10,56 @@ const MafiaGame = require('../domain/MafiaGame');
...
+  });
 
-        // Subscribe to the room events
-        socket.join(roomID);
+    // this overlaps with justins code.
```
Comment (UltimaKuma) [inline backend/Events/LobbyEvents.js:30]: Had a talk with @Dhors , this section of code will most likely be overwritten by @jteo97's PR since it overlaps.

File: backend/package-lock.json [hunk 3]
```diff
@@ -1,8336 +1,8 @@
 {
     "name": "mafia-backend",
     "version": "0.1.0",
-    "lockfileVersion": 2,
+    "lockfileVersion": 1,
     "requires": true,
-    "packages": {
```
Comment (UltimaKuma) [inline backend/package-lock.json:6]: Why was this deleted?

File: backend/test/index.test.js [hunk 4]
```diff
@@ -1,85 +1,127 @@
...
+
+    // todo: doesnt work until justin pushes his code in.
+    // // Or Client should recieve a to game-ready - this is the non-host client.
+    // clientSocket.on("game-ready", () => {
+    //     // Host should receive a different event than the client
+    //     done();
+    // });
+
+    // Request to server that lobby is full.
+    clientSocket.emit("lobby-ready", createLobbyDTO);
```
Comment (UltimaKuma) [inline backend/test/index.test.js:106]: Think the server should emit to the client that the lobby is ready such that the client can enable the button to start the game. 

Under this implementation it is the client that is telling the server that the lobby is ready which I think is incorrect

File: backend/test/index.test.js [hunk 5]
```diff
@@ -1,85 +1,127 @@
...
-            done();
-        });
+    // Request to create a new lobby
+    // this will set the socket id to the host to be tested.
+    clientSocket.emit("create-lobby", createLobbyDTO);
 
-        // Request to create a new lobby
-        clientSocket.emit('create-lobby', createLobbyDTO);
+    // Host client should subscribe to confirm-game-start - this is the host client.
+    clientSocket.on("confirm-game-start", () => {
```
Comment (UltimaKuma) [inline backend/test/index.test.js:93]: "confirm-game-start" was not an agreed upon endpoint for the API.

**Review by UltimaKuma — CHANGES_REQUESTED** [review]
Really like how you've created the domain model tests. However I have a few questions and comments that I would like resolved.

**Review by parfei — CHANGES_REQUESTED** [review]
Nice job Daniel! I will come back and review some more :)

File: backend/Events/LobbyEvents.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (parfei) [reply backend/Events/LobbyEvents.js:30]: Just saw Leon's comment.

Comment (parfei) [reply backend/Events/LobbyEvents.js:30]: @Dhors a nitpick but can you append TODO in the comment?

File: backend/Events/LobbyEvents.js [hunk 6]
```diff
@@ -10,23 +10,56 @@ const MafiaGame = require('../domain/MafiaGame');
...
+      // host is position 0, so skip.
+      for (i = 1; i < mafiaGamePlayers.length; i++) {
+        let player = mafiaGamePlayers[i];
+        io.to(hostsocketID).emit("game-ready");
+      }
     });
-}
+
+    // host has clicked start game
+    socket.on("start-game", () => {
```
[the remark this answers is not in the document]
Comment (Dhors) [reply backend/Events/LobbyEvents.js:52]: sree does game-start
very different :)

File: backend/test/index.test.js (hunk 4 above)
Comment (Dhors) [reply backend/test/index.test.js:106]: i deleted the event handler

File: backend/test/index.test.js (hunk 5 above)
Comment (Dhors) [reply backend/test/index.test.js:93]: i deleted the event handler

File: backend/Events/LobbyEvents.js [hunk 7]
```diff
@@ -10,23 +10,56 @@ const MafiaGame = require('../domain/MafiaGame');
...
+    socket.on("lobby-ready", () => {
+      let roomID = socket.player.roomID;
 
-        // Add player information to the host socket
-        socket.player = host;
+      let mafiaGameRoom = mafiaGame.gameRoomsDict[roomID];
 
-        // Send room ID back to host.
-        io.in(roomID).emit('lobby-code', new LobbyCodeDTO(roomID));
+      let hostsocketID = mafiaGameRoom.players[0].socketID; //(host is first person)
```
[the remark this answers is not in the document]
Comment (Dhors) [reply backend/Events/LobbyEvents.js]: so a host field? I can add it i guess, but do we need it?
host is only ever really used in this instance, and its only useful for the case that if the host disconnects (which we arent really covering right)

File: backend/Events/LobbyEvents.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (UltimaKuma) [reply backend/Events/LobbyEvents.js:32]: Think sree was doing both, this will most likely be overwitten by his PR

**Review by UltimaKuma — APPROVED** [review]
LGTM, nice work on the tests and bug fix.

File: backend/Events/LobbyEvents.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (Dhors) [reply backend/Events/LobbyEvents.js]: added

File: backend/domain/Room.js [hunk 8]
```diff
@@ -11,7 +11,7 @@ class Room {
         // Default game settings.
         this.gameState = INITIAL_GAME_STATE;
         this.maxPlayerCount = 6;
-        this.players = new Array(this.maxPlayerCount);
+        this.players = new Array();
```
[the remark this answers is not in the document]
Comment (Dhors) [reply backend/domain/Room.js:14]: ty

File: backend/test/domainmodel.test.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    let room = new Room();
+
+    room.addPlayer(player1);
+    room.addPlayer(player2);
+    room.addPlayer(player3);
+    room.addPlayer(player4);
+    room.addPlayer(player5);
+    room.addPlayer(player6);
+
+    if (room.players.length === 6) {
```
Comment (parfei) [inline backend/test/domainmodel.test.js:24]: Nitpick: It might be better if we use a expect() matcher here - would be more clearer! Same as in the other tests.

File: backend/test/domainmodel.test.js [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    }
+  });
+
+  test("test room domain model 1 object", (done) => {
+    let player1 = new Player("00001", "room1", "nickname1");
+
+    let room = new Room();
+
+    room.addPlayer(player1);
+
```
Comment (parfei) [inline backend/test/domainmodel.test.js:35]: Can you also test edge cases like no players added? You can probably add them in a different PR tbh, not too high of a priority I think..

**Review by parfei — CHANGES_REQUESTED** [review]
Nice changes Daniel :), appreciate your hard work! Love that you added tests for the bug fix. Btw, I don't think you're supposed to delete the `package-lock.json`. Also, should the file be named DomainModel.test.js? Not sure.

**Review by parfei — APPROVED** [review]
Nice - had a chat with Daniel just now. The implementation of `confirm-start` event handler will be done by @jteo97.

`game-start` will be overridden by @SreenikethRaghavan 's changes.

This PR mainly fixes the mentioned bug.
